### PR TITLE
Remove requirement that GCAllocator must be shared

### PR DESCRIPTION
**Pull request by schveiguy**
There is no internal state in `GCAllocator`, it just uses the state of the GC via global variable access.

Requiring one to use `shared(GCAllocator)` everywhere is annoying. Often times one wants to have an allocator inside an object or otherwise, and it sucks to have to mark it shared.

This change allows access via shared OR unshared `GCAllocator`.

I also saw an inefficiency in `resolveInternalPointer`. I can create a separate pull if this is controversial.

Flagging @andralex, this is somewhat of a design decision, even though the change is transparent.

### CODE REVIEW

File: std/experimental/allocator/gc_allocator.d [hunk 1]
```diff
@@ -72,22 +72,22 @@ struct GCAllocator
     }
 
     /// Ditto
-    void[] resolveInternalPointer(void* p) shared
+    static void[] resolveInternalPointer(void* p)
```
Comment (JackStouffer) [inline std/experimental/allocator/gc_allocator.d]: No reason this can't be `@trusted` to, right?

**Comment by JackStouffer** [comment]
I don't see anything in the docs saying `query` is faster than `addrOf`. Have you done any profiling on `resolveInternalPointer `?

**Comment by schveiguy** [comment]
No I haven't. I just know how it's implemented. Addrof gets the block info and returns the pointer. Sizeof gets the block info and returns the size. Query returns the block info :)

File: std/experimental/allocator/gc_allocator.d (hunk 1 above)
Comment (schveiguy) [inline std/experimental/allocator/gc_allocator.d]: I don't see why not.

**Comment by JackStouffer** [comment]
Ping @schveiguy, still has the nitpick

**Comment by schveiguy** [comment]
@JackStouffer sorry, forgot about this. Updated.

**Comment by JackStouffer** [comment]
LGTM

**Comment by ZombineDev** [comment]
I think we should also get rid of the Allocator.instance idiom which breaks purity, but we should leave that discussion for another PR. LGTM

**Comment by schveiguy** [comment]
@ZombineDev This shouldn't break purity, instance is not needed for any of the calls. In fact, the allocator in general needs a bunch of attribute TLC (`nothrow` should be everywhere, `pure`, `@safe`/`@trusted`).

**Comment by schveiguy** [comment]
I tested, everything can be made `pure nothrow` except for `collect`.

I'm surprised `GC.free` is `pure`...

**Comment by wilzbach** [comment]
> I'm surprised GC.free is pure...
> Not going to change this PR, because it has a different focus. That could be done in another PR.

I already tried that see e.g. #4190 or #4269 - however there's still hope:

https://github.com/dlang/phobos/pull/3891

**Comment by ZombineDev** [comment]
@schveiguy Yes none of the calls to the GC now come through the `instance` instance methods, however consider the case GCAllocator is the parent allocator of another allocator.
Correct me if I'm wrong, but then just because the new sub-allocator has a shared field, it becomes unusable in pure code, regardless if the its methods are implemented by calling the static pure functions of GCAllocator.
Also, `make` would not be inferred as pure, because one of its arguments is `shared`.

The correct idiom IMO is to change all functions that are templated on the allocator type (excluding those that work with `IAllocator`) from:

```
void useAllocator(A)(auto ref A allocaor)
{
    void[] mem = allocator.allocate(42);
}

useAllocator(GC.instance);
// useAllocator(GC); doesn't compile
```

to:

```
void useAllocator(alias allocator = GCAllocator)()
{
    // if the argument is a type with static methods, getAllocatorInstance 
    // aliases itself to the type (allocator), if there are no static methods - it aliases to
    // Type.instance (allocator.instance). Else if allocator is an allocator instance it
    // just aliases to it.
    alias alloc = getAllocatorInstance!allocator;
    void[] mem = allocator.allocate(42);
}

useAllocator() // uses GCAllocator static methods
useAllocator!Mallocator() // uses  Mallocator.instance instance methods
useAllocator!(Mallocator.instance)() //works with instances too
useAllocator!theAllocator()
```

That way getAllocatorInstance abstracts the fact if you're calling static or instance methods. Also it makes it easier to provide a default parameter value. For example:

```
// if you change the ddeclaration from:
void useAllocator(alias allocator = GCAllocator)()
// to
void useAllocator(alias allocator = processAllocator)()
// you don't need to change the signature (template type parameter default to runtime parameter default) or the implementation of the function.
```

**Comment by schveiguy** [comment]
> but then just because the new sub-allocator has a shared field, it becomes unusable in pure code

The shared field is static, so it's not a member of any instance, it's more like a global variable. Yes, the `instance` is marked shared, but can be cast implicitly to unshared since it has no shared data:

```D
GCAllocator gcalloc = GCAllocator.instance;
```

This can be used to compose other allocators, no problem, without the `shared` tag.

**Comment by ZombineDev** [comment]
@schveiguy about the `pure` `GC.free`. I have been thinking about it and it is actually pure, depending on how you look at it. Consider a local container variable inside a pure function. Since you're the only one knowing about it, inserting and removing elements from it is ok. Now consider that allocators are just some kind of containers. If only you know about it, adding and removing elements from them is ok (allocating and deallocating). If you're the only one with a reference to a piece of memorying (given good enough allocator implementation -  e.g. everything is thread-local) then the effect of freeing this memory should not be visible anywhere else in the program and so the illusion of purity should remain well.

**Comment by schveiguy** [comment]
> it is actually pure

The trouble with making `GC.free` pure, is that the compiler isn't aware that you are accessing global data.

I imagine that this kind of scenario can happen:

```D
void freeIt(S)(S *ptr)
{
   GCAllocator.deallocate(cast(void *) ptr);
}
```

Seems innocuous enough, right? But what if `S` is `immutable(int)`? Due to `GCAllocator.deallocate` being pure, the compiler infers that `freeIt` is also pure. Since `ptr` has no non-immutable indirections, this instantiation of `freeIt` can be assumed to be *strong* pure. And as such, a strong pure function that returns void **doesn't need to be called**!

In terms of GC allocations, this isn't the end of the world, since the GC would (should?) come along and collect it anyway. But the idea that this isn't actually called bugs me.

**Comment by ZombineDev** [comment]
But, AFAIR in practice child allocators save the parent allocator in an instance field and can't take advantage of the fact that they can use the global variable directly instead of saving a reference to it.

**Comment by schveiguy** [comment]
> But, AFAIR in practice child allocators save the parent allocator in an instance field and can't take advantage of the fact that they can use the global variable directly instead of saving a reference to it.

I'm not understanding your argument, sorry.

**Comment by schveiguy** [comment]
Heh, I could do this instead:

```D
alias instance = typeof(this).init;
```

**Comment by ZombineDev** [comment]
Yes I'm aware of this problem. I think we need some sort of `@weakpure` transitive attribute that allows calling such functions from `pure` code, but prevent the compiler from eliminating the calls.

**Comment by schveiguy** [comment]
The `@weakpure` solution doesn't work. If you can wrap it in `pure`, then you can cause the problem.

**Comment by ZombineDev** [comment]
You forget that allocators don't need to be a global static (state-less) resource. I use them extensively as statefull local variables and then the metaphore with containers works nicely.

> Heh, I could do this instead

This exactly what my getAllocatorInstance does. Just abstracts the fact if you're calling static or instance methods.

**Comment by schveiguy** [comment]
@ZombineDev I'm still not getting it. Why can't you use `GCAllocator` in some type?

```D
struct S(Allocator)
{
   Allocator allocator;
   auto foo() pure { allocator.allocate(100); }
}

auto s = S!GCAllocator(GCAllocator.instance); // should be fine
```

**Comment by ZombineDev** [comment]
I imagine it like this

int compute() pure
{
     allocate() weak pure
     // do some stuff
     free() weak pure
}

So there's no problem if compute() call is elided, because the weakpure calls come in pair.

**Comment by ZombineDev** [comment]
Also note that allocate and deallocate are `@system`.  High-level pure code just uses UniqueRef/SharedRef which provide `@safe` mem allocation interfacd.

**Comment by schveiguy** [comment]
> So there's no problem if compute() call is elided

Sure, but `GC.free` is already weak pure (it accepts a mutable pointer), that isn't the issue.



> because the weakpure calls come in pair.

How is that enforced?

**Comment by ZombineDev** [comment]
It will be enforced by safe or trusted wrappers. User code shouldn't call those system functions on its own.

**Comment by ZombineDev** [comment]
> I'm still not getting it. Why can't you use GCAllocator in some type?

1) I didn't think this would work without casts:

```
GCAllocator gcalloc = GCAllocator.instance;
```

2) A singleton with no instance methods and state is useless.

3) using alias + getAllocatorInstance is better because you wouldn't need to have another field embedded in your object, your code would not care if there's an instance or if it's calling static methods and it workarounds the shared problems in case 1) doesn't always work.

**Comment by schveiguy** [comment]
> I didn't think this would work without casts:

Should be fine, since no instance data is a reference.

> A singleton with no instance methods and state is useless.

I can't really get rid of the instance, since existing code may expect lvalue access. This is experimental, so we could potentially alter this, but I'd rather not put too much into this PR.

> using alias + getAllocatorInstance is better because you wouldn't need to have another field embedded in your object

I'm assuming you mean as a user of allocators? I don't see why you can't do this. Just substitute `instance` for `getAllocatorInstance`.

**Comment by ZombineDev** [comment]
> I'm still not getting it. Why can't you use GCAllocator in some type?

1) I didn't think this would work without casts:

```
GCAllocator gcalloc = GCAllocator.instance;
```

2) A singleton with no instance methods and state is useless.

3) using alias + getAllocatorInstance is better because you wouldn't need to have another field embedded in your object, your code would not care if there's an instance or if it's calling static methods and it workarounds the shared problems in case 1) doesn't always work.

**Comment by andralex** [comment]
This is by design. What is the situation in which you need to instantiate `GCAllocator`?

**Comment by schveiguy** [comment]
The impetus for this is because I want to handle in my code the ability to use any kind of allocator -- ones that are instantiated, and ones that are singletons. I have code like this:
```D
struct Foo(Allocator)
{
   Allocator alloc;
}
```
I found that `Foo!GCAllocator` doesn't work, because `alloc` needs to be marked `shared` in order to be used (yes, I can do `Foo!(shared(GCAllocator))`, but it doesn't make sense, as I'm not going to share `alloc`).

But the requirement rings untrue, none of the functions actually require shared state, because all the state is external. I realize, I can put in specialized code to deal with the two forms, but I wanted something simple. This is what people are going to reach for if writing generic code.

I don't see how making `GCAllocator` functions static harms anything, it just allows more possibilities for usage without affecting existing code.

I just looked at other usages of singletons in your wrapper types and see that you use an alias if the allocator has no state. Is there a way we can make this easy to use for users (maybe a mixin)? Then I can drop this PR, and create the proper one. This can give people who are looking for the right solution the proper way to do so.

**Comment by schveiguy** [comment]
Also, using `GCAllocator.instance` is impure, even though it's really pure. How do you solve that problem?

**Comment by schveiguy** [comment]
I can think of 3 cases that a generic type needs to handle:

1. The allocator is a shared singleton
2. The allocator can be default constructed, and you just use it.
3. The allocator is passed into the type/function and should be used.

We need some good documentation on how to handle all 3 cases generically. Are there more than 3?

**Comment by ZombineDev** [comment]
@andralex if it wasn't clear from the discussion above, I agree with Steven that `GCAllocator`'s methods should be static and also not `shared`, as current design is much less useful compared  to `core.memory`, while I think we all agree that it should be the other way around. The bare minimum being that `GCAllocator` should be usable in `pure` code, which currently is not the case (otherwise what's advantage over just using `new`?).

Just a quick note about `shared` - `shared` instance methods only make sense when they are overloaded with non-shared methods in order to provide a different behavior (e.g. synchronization).
When there is A) no state, or B) when the state being modified is external to the object (and is internally synchronized), `shared` doesn't buy us anything and is just a usability pessimization.
That applies to `Mallocator`, as well as `GCAllocator`. `MmapAllocator` may need some synchronization, beyond what the OS gives us, to allow concurrent allocation, so that's one instance where `shared` makes sense.

The other issue that is not very clear, is how should generic code use allocators. There are several alternatives:
1. Use `processAllocator` and/or `theAllocator`. While this may be an OK default, very often it's more efficient to use a dedicated bump-the-pointer (may be even placed on the stack) allocator for short-lived allocations. That's why generic code should allow the user to override the default allocator whenever possible. Plus, using global variables is not possible in `pure` code, and also messing with global state may be dangerous.
2. Take an `IAllocator` parameter. While this is bit more flexible than 1) (plus users coming from languages like Java will feel right at home), this misses one the main advantage of the allocator design - i.e. design by introspection. Furthermore, classes and interfaces are still quite cumbersome to use in `@nogc` code, though `allocatorObject` can help.
3. Take an `auto ref A allocator` parameter. This has the benefit of allowing design by introspection, however has two other disadvantages compared to 1) and 2). The first is that aggregates (e.g. data structures or child allocators) can't save a reference for later use, because in D `ref` is not a type constructor like `T&` in C++, which translates to the fact that ref fields are not allowed, and instead the constructor will probably make a copy if the allocator is a value type. This is highly undesirable when the allocator is stateful and you actually want to share the state (e.g. so you can do `deallocateAll` when you're done, and not leak memory because the data structure is actually using a copy of the allocator).
The other disadvantage is that you can't easily provide default runtime or template type arguments, like `processAllocator` or `GCAllocator`. This not good because it is desirable to provide a safe default (such as `GCAllocator`) for high-level code, eg. scientific applications that use `ndslice.makeSlice`, but doesn't preclude customization down the road.
4. Use `alias allocator` template parameter. This pattern is not used anywhere in Phobos (yet), but I found it very useful in my projects and that's why I'm proposing it here. The advantages are:
* allows design by introspection (i.e. doesn't do type erasure, like IAllocotor), so it's easy to specialize based on the allocator. For example, I have a `SharedRef(T, alias allocator)` type (similar to Phobos `RefCounted` and C++ `shared_ptr`) which if the allocator has a `prefix` or `suffix` member uses one of them for storing the ref count, otherwise, manges the RC heap allocation manually.
* Allows saving a reference to the allocator, even if the allocator is a local variable living entirely on the stack. For example, the only field that my `UniqueRef` type has is a pointer to the payload. I would expect that if the compiler can see that an instance of `UniqueRef` doesn't outlive the stack frame where the stack allocator lives it would be able to figure out that closure context is not necessary and replace all references to the allocator with a stack pointer offset, making `UniqueRef.sizeof == void*.sizeof` and thereby making my `UniqueRef` a truly zero-const-abstraction, which in turn makes `@safe @nogc pure` possible and actually pleasant to write.
+ Makes default parameters easy and flexible:

  ```
  auto makeArray(alias allocator = GCAllocator, Range)(Range elementsToCopy); 
  // or:
  auto makeArray(alias allocator = theAllocator, Range)(Range elementsToCopy); 
  ```

**Comment by JackStouffer** [comment]
Ping @andralex

**Comment by andralex** [comment]
I don't think generic code should instantiate allocators such as `GCAllocator`. The reality is there is only one garbage collected allocator, it's shared by all threads, so modeling it as such seems like a sensible thing to do.

**Comment by andralex** [comment]
Furthermore, the `shared` attribute is useful in ensuring that code like freelists (thread-local or shared) on top of `GCAllocator` are sensible. No?

**Comment by wilzbach** [comment]
> unusable in pure code, regardless if the its methods are implemented by calling the static pure functions of GCAllocator.
> Also, make would not be inferred as pure, because one of its arguments is shared.
..
> @andralex if it wasn't clear from the discussion above, I agree with Steven that GCAllocator's methods should be static and also not shared, as current design is much less useful 

FWIW I couldn't make the unittests of `make` and `makeArray` `pure`, because of this limitation (shared -> impure), which is of course annoying for users of an API. After all they expect the same behavior of the GCAllocator `makeArray!int(10)` compared to the native one `new int[10]` and if @andralex's plans are true we want to unify these two anyways soon :)

**Comment by schveiguy** [comment]
I'll close for now. It's not a critical requirement for me as I had to write my own GC wrapper since this one doesn't allow me to avoid scanning for pointers in my buffer library. No sense in consuming auto-tester cycles.

> The reality is there is only one garbage collected allocator, it's shared by all threads, so modeling it as such seems like a sensible thing to do.

The shared instance doesn't go away here (existing code should just work). It's just that the calls would now be pure, and reflect more accurately the actual model -- that there isn't an accessible "instance" of the GC.

> I don't think generic code should instantiate allocators such as `GCAllocator`.

I hope this isn't a serious reason why `GCAllocator` methods are marked `shared`. There are much better ways to ensure you can't instantiate a local instance of something.

> Furthermore, the shared attribute is useful in ensuring that code like freelists (thread-local or shared) on top of `GCAllocator` are sensible. No?

I don't get the point of this comment.

Note: we still need a guide on how generic code should deal with all the models of allocators (shared instance, local constructed instance, passed instance, etc.). Until that, it's hard to have a discussion on what the API should be.

**Pull request closed without merging**